application/frontend/controllers/AuthController.php

```diff
@@ -0,0 +1,123 @@
+            // @todo can this be changed to use a URL parameter for the
+            // token for this action and we can pass to idp and back?
+            $this->enableCsrfValidation = false;
+        }
+        return parent::beforeAction($action);
+    }
+
+    public function actionLogin()
+    {
+        /*
+         * Collect return to url of where to send user after successful login
+         * Expected as relative url starting with /
+         * Before redirecting user after login this will be prefixed with ui_url
+         */
+        $returnTo = \Yii::$app->request->get('ReturnTo', '');
+
+        if ( ! \Yii::$app->user->isGuest) {
+            $afterLogin = $this->getAfterLoginUrl($returnTo);
+            return $this->redirect($afterLogin);
+        }
+
+        /*
+         * Initialize $log variable for logging
+         */
+        $log = ['action' => 'login'];
+
+        try {
+            $returnTo = Url::to(['auth/login', 'ReturnTo' => $returnTo], true);
+            /** @var AuthUser $authUser */
+            $authUser = \Yii::$app->auth->login(\Yii::$app->request, $returnTo);
+
+            $log['email'] = $authUser->email;
+            // Get local user instance or create one
+            $user = User::findOrCreate(null, null, $authUser->employeeId);
+            // Initialize session for user
+            if (\Yii::$app->user->login($user, \Yii::$app->params['sessionDuration'])) {
+                $log['status'] = 'success';
+                \Yii::warning($log, 'application');
+
+                $afterLogin = $this->getAfterLoginUrl($returnTo);
+                return $this->redirect($afterLogin);
+            } else {
+                throw new UnauthorizedHttpException('Unable to perform user login', 1459966846);
+            }
+        } catch (RedirectException $e) {
+            /*
+             * Login triggered redirect to IdP to login, so return a redirect to it
+             */
+            return $this->redirect($e->getUrl());
+        } catch (\Exception $e) {
+            $log['status'] = 'error';
+            $log['error'] = $e->getMessage();
+            $log['code'] = $e->getCode();
+            \Yii::error($log, 'application');
+            throw new UnauthorizedHttpException($e->getMessage(), $e->getCode());
+        }
+
+    }
+
+    public function actionLogout()
+    {
+        if ( ! \Yii::$app->user->isGuest) {
+            $authUser = \Yii::$app->user->identity->getAuthUser();
+        } else {
+            /*
+             * User not logged in, but lets kill session anyway and redirect to UI
+             */
+            \Yii::$app->user->logout(true);
+            return $this->redirect(\Yii::$app->params['ui_url']);
+        }
+
+        /*
+         * Kill local session
+         */
+        \Yii::$app->user->logout(true);
+
+        /*
+         * Log user out of IdP
+         */
+        try {
+            \Yii::$app->auth->logout($authUser, \Yii::$app->params['ui_url']);
```

Comment:
If the call to `\Yii::$app->auth->logout(...)` doesn't throw a `RedirectException`, what would happen?  Should we instead just be getting the URL from the `RedirectException` (if thrown, otherwise using the ui_url), and then waiting until after the `catch` block to do the redirect?